src/MongoDB.Driver/Linq/Processors/ExpressionValidator.cs

```diff
@@ -0,0 +1,122 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to validate a LINQ expression.
+    /// </summary>
+    internal sealed class ExpressionValidator : ExpressionVisitor
```

Comment:
This feels a lot more complicated than what we had discussed the other day.  What were the expressions that failed that much simpler form?

Reply:
Yep, I've updated the logic because this code with a query to local collection, should not throw the exception:
```
      var local = new List<string> { "itchy" };
      Assert(
                x => local.Any(i => x.C.E.I.Contains(i)),
                1,
                "{\"C.E.I\": { $in: [\"itchy\" ] } }");
```

Reply:
Feels like you could just skip the check if the `Object` of the `MethodCallExpression`, Any in this case, was rooted at a constant.

Reply:
as I see, this is exactly what I've done:

                    var memberExpression = argument as MemberExpression;
                    if (memberExpression != null)
                    {
                        // add top-level expression parameters to the list
                        // with parent's parameters.
                        // NOTE: these parameters will be added to all members,
                        // even if the member is not in the top-level, but since
                        // top-level parameters should not be placed in any level 
                        // except of direct children, it's not a problem
                        AddTopLevelExpressionParameters(_inParametersScope);
                        continue;
                    }
here there is only logic to add the expression parameters from top-level expression, because when we call `MethodCallExpression` first time, we don't have expression parameters values in input node.

Reply:
I agree with Craig that this method feels too complicated. I can't completely figure out what it's doing...